### ISSUE TITLE
Envoy OAuth2 Filter TypeAssertionErrors

### ISSUE THREAD

**Issue by rijos** [istio/istio#36537]
### Bug Description

Adding the following filter to the filterchain results in typecasting errors for istiod:

```yaml
kind: EnvoyFilter
...
  - applyTo: HTTP_FILTER
    match:
      listener:
        portNumber: 8080
        filterChain:
          filter:
            name: "envoy.filters.network.http_connection_manager"
            subFilter:
              name: "envoy.filters.http.oauth2"
    patch:
      operation: INSERT_BEFORE
      value:
        name: envoy.filters.http.oauth2
        typed_config:
          "@type": type.googleapis.com/envoy.extensions.filters.http.oauth2.v3.OAuth2
          config:
            does: not matter
```

Error in istiod:

```
error   Observed a panic: &runtime.TypeAssertionError{_interface:(*runtime._type)(0x28d7da0), concrete:(*runtime._type)(nil), asserted:(*runtime._type)(0x2bf2aa0), missingMethod:""} (interface conversion: protoreflect.ProtoMessage is nil, not *envoy_extensions_filters_network_http_connection_manager_v3.HttpFilter)
goroutine 29635 [running]:
istio.io/istio/pilot/pkg/util/runtime.LogPanic({0x28be620, 0xc0020c5a70})
        istio.io/istio/pilot/pkg/util/runtime/runtime.go:29 +0x4d
istio.io/istio/pilot/pkg/util/runtime.HandleCrash({0xc00130e1d0, 0x2, 0x7ff20f9aa968})
        istio.io/istio/pilot/pkg/util/runtime/runtime.go:37 +0x5b
panic({0x28be620, 0xc0020c5a70})
        runtime/panic.go:1038 +0x215
istio.io/istio/pilot/pkg/networking/core/v1alpha3/envoyfilter.patchHTTPFilters(0x3, 0xc001c43580, 0x2dc7004, 0x2f0c640, 0xc0023107d0)
```

Changing the typed_config to "@type": type.googleapis.com/envoy.extensions.filters.http.oauth2.v3alpha.OAuth2
enables istiod to push it's changes to istio-ingress, only to be met by the following error:

istio-ingress:
```
[libprotobuf WARNING external/com_google_protobuf/src/google/protobuf/text_format.cc:2065] Can't print proto content: 2021-12-16T15:23:31.547547Z     warning envoy config    gRPC config for type.googleapis.com/envoy.config.listener.v3.Listener rejected: Error adding/updating listener(s) 0.0.0.0_8080: Unable to unpack as envoy.extensions.filters.http.oauth2.v3.OAuth2: type_url: "type.googleapis.com/envoy.extensions.filters.http.oauth2.v3alpha.OAuth2"
value: "\n\374\001\n\034\n\017oauth.com/token\032\002\010\003\022\005oauth\022\"https://oauth.com/oauth/authorize/\032N\n\003foo\022\'\n\005token\022\036\n\034/etc/envoy/token-secret.yaml\032\036\n\004hmac\022\026\n\024/etc/envoy/hmac.yaml\"6%REQ(:x-forwarded-proto)%://%REQ(:authority)%/callback*\r\n\013\n\t/callback2\014\n\n\n\010/signoutJ\004userJ\006openidJ\005email"
```

To make sure the required fields are present we simply copy-pasted the typed_config from this sample (with the exception of @type): https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/oauth2_filter

```yaml
  - applyTo: HTTP_FILTER
    match:
      listener:
        portNumber: 8080
        filterChain:
          filter:
            name: "envoy.filters.network.http_connection_manager"
            subFilter:
              name: "envoy.filters.http.oauth2"
    patch:
      operation: INSERT_BEFORE
      value:
        name: envoy.filters.http.oauth2
        typed_config:
          "@type": type.googleapis.com/envoy.extensions.filters.http.oauth2.v3alpha.OAuth2
          config:
            token_endpoint:
              cluster: oauth
              uri: oauth.com/token
              timeout: 3s
            authorization_endpoint: https://oauth.com/oauth/authorize/
            redirect_uri: "%REQ(:x-forwarded-proto)%://%REQ(:authority)%/callback"
            redirect_path_matcher:
              path:
                exact: /callback
            signout_path:
              path:
                exact: /signout
            credentials:
              client_id: foo
              token_secret:
                name: token
                sds_config:
                  path: "/etc/envoy/token-secret.yaml"
              hmac_secret:
                name: hmac
                sds_config:
                  path: "/etc/envoy/hmac.yaml"
            # (Optional): defaults to 'user' scope if not provided
            auth_scopes:
            - user
            - openid
            - email
 ```
Looks like types between istio and envoy dont match?

### Version

```prose
client version: 1.12.1
control plane version: 1.12.1
data plane version: 1.12.1 (3 proxies)
envoy  version: e6f45abcf874983fbff384459d70b28c072f68b5/1.20.1/Clean/RELEASE/BoringSSL
```

**Comment by howardjohn** [comment]
Guessing this is due to https://www.envoyproxy.io/docs/envoy/latest/version_history/v1.20.0#incompatible-behavior-changes first entry

**Comment by pedromarceOAT** [comment]
@howardjohn is there any expectation for the release date for this fix in release 1.12?

**Comment by howardjohn** [comment]
Fixed by https://github.com/istio/istio/pull/36577. Will be in 1.12.2 released this week
